presto-main/src/test/java/io/prestosql/sql/query/TestJoin.java

```diff
@@ -52,4 +53,54 @@ public void testCrossJoinEliminationWithOuterJoin()
                         "JOIN d ON d.id = a.id"))
                 .matches("VALUES 1");
     }
+
+    @Test
+    public void testInPredicateInJoinCriteria()
+    {
+        assertThat(assertions.query("SELECT * FROM (VALUES 1, 2, NULL) t(x) JOIN (VALUES 1, 3, NULL) u(x) ON t.x IN (VALUES 1)"))
+                .matches("VALUES (1, 1), (1, 3), (1, NULL)");
+
+        assertThat(assertions.query("SELECT * FROM (VALUES 1, 2, NULL) t(x) LEFT JOIN (VALUES 1, 3, NULL) u(x) ON t.x IN (VALUES 1)"))
+                .matches("VALUES (1, 1), (1, 3), (1, NULL), (2, NULL), (NULL, NULL)");
+
+        assertThat(assertions.query(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) RIGHT JOIN (VALUES 1, 3, NULL) u(x) ON t.x IN (VALUES 1)"))
+                .matches("VALUES (1, 1), (1, 3), (1, NULL)");
+
+        assertThat(assertions.query(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) FULL JOIN (VALUES 1, 3, NULL) u(x) ON t.x IN (VALUES 1)"))
+                .matches("VALUES (1, 1), (1, 3), (1, NULL), (2, NULL), (NULL, NULL)");
+
+        assertThat(assertions.query(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) JOIN (VALUES 1, 3, NULL) u(x) ON u.x IN (VALUES 1)"))
+                .matches("VALUES (1, 1), (2, 1), (NULL, 1)");
+
+        assertThat(assertions.query(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) LEFT JOIN (VALUES 1, 3, NULL) u(x) ON u.x IN (VALUES 1)"))
+                .matches("VALUES (1, 1), (2, 1), (NULL, 1)");
+
+        assertThat(assertions.query(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) RIGHT JOIN (VALUES 1, 3, NULL) u(x) ON u.x IN (VALUES 1)"))
+                .matches("VALUES (1, 1), (2, 1), (NULL, 1), (NULL, 3), (NULL, NULL)");
+
+        assertThat(assertions.query(
+                "SELECT * FROM (VALUES 1, 2, NULL) t(x) FULL JOIN (VALUES 1, 3, NULL) u(x) ON u.x IN (VALUES 1)"))
```

Comment:
What if IN contains correlated subquery? Can you please test that (with symbols from left and right relations)?

Reply:
That's not supported. It fails with an error saying that the reference is not allowed in that context: https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java#L1501. I'll add some tests.